### ISSUE TITLE
Quick Launch Tasks Collapse Regions no longer works

### ISSUE THREAD

**Issue by lazworks** [Microsoft/VS-PPT#158]
Collapse Regions no longer works after the latest updates to Visual Studio 2017. It was working before.

Create one or more regions in code.
Activate Quick Launch and type Col and then click on Collapse Regions.
No action is taken.

**Comment by ygoe** [comment]
ExpandRegions also stopped working. This extension is now officially useless. Too bad. There are no alternatives.

**Comment by ygoe** [comment]
I can't find any code related to this feature in the published source code. So this repository clearly can't be the source for the published extension. This extension is NOT open-source!

**Comment by olegtk** [comment]
Yes, Quick Launch tasks extension has not been open sourced yet. It uses some internal API that we wanted to replace before publishing it on this repo.
Unfortunately Collapse/Expand Regions is implemented in a bit of a hacky way and recent Roslyn changes just broke it :(

**Comment by justcla** [comment]
Just to be clear, the extension is not officially useless. It's not official until we say so. :-P
And in the meantime, there are plenty of officially useful features in QuickLaunch tasks, including PresentOn and PresentOff.

**Comment by ygoe** [comment]
Yeah, sorry, it's not officially useless, that's just for me. And if Roslyn internals broke the extension's hacks, I'd probably not find a solution in the code anyway.

**Comment by justcla** [comment]
There are potential fixes. We've been looking into them. It's mostly about changing the way we detect a region.
It's looking like we'll be able to open-source the code for Quick Launch tasks. Hopefully we'll have it out by Thanksgiving.

**Comment by michaelcsikos** [comment]
_CollapseRegions_ is the only command I use from **Quick Launch Tasks**. I came here looking for the source code because I wanted to make a new extension that allowed this action to be assigned a shortcut. I type **Ctrl+Q, c, o, l, Enter** a hundred times a day. Six keystrokes every time! When you fix it, could you allow a shortcut to be assigned, please? Retrofitting it to VS 2013 an 2015 would be great, too.

**Comment by justcla** [comment]
Alright, I've woken @gundermanc. He's working on it now. :-)
If he gets a fix in, I'll package it, and send it out for testing.

**Comment by gundermanc** [comment]
Ok, we have a confirmed fix. Always a pleasure! @justcla will work on getting it packaged.

**Comment by ygoe** [comment]
@michaelcsikos Great idea to assign it an even shorter hotkey. I use Collapse/ExpandRegions quite often, too, and all that typing (and the waiting for UI synchronisation/feedback) adds up.

The new quick launch tasks would probably need to be registered as actions like all other menu commands so the hotkey manager can be used as usual.

**Comment by justcla** [comment]
Agreed. Shortcuts would make it much more useful.
And yes, we'd need to define some commands in a VSCT file, then create handlers for the commands.
Then users can hook up shortcuts to the commands.

That's a fair bit of work. I don't think I'm likely to get around to it this year.

**Comment by justcla** [comment]
As for updating VS2013 and VS2015, we don't touch those extensions anymore. Too much risk. Instead, we recommend people upgrade to VS2017.

**Comment by michaelcsikos** [comment]
Some projects specify which version of Visual Studio we must use. For AutoCAD 2018 add-ins, we must Visual Studio 2015 Update 3, according to Autodesk.

If only the _CollapseRegions_ was open source, we could make a new use-at-our-yown-risk add-in with shortcut support...

**Comment by justcla** [comment]
We are working on open-sourcing Quick Launch Tasks. Check with @prnadago for an update. :-)

**Comment by prnadago** [comment]
Yes it's on my backlog! But it still might be early in the new year when the work gets done.

In the meantime. expand/collapse regions is fixed! You can get the updated vsix from - https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.QuickLaunchTasks#overview

This issue can be closed now :)

**Comment by lazworks** [comment]
Thanks prnadago.
Unfortunately, after downloading the update, the behaviour is unchanged (ie Collapse All Regions does not work as described in my original post).

**Comment by prnadago** [comment]
Hmm that's unfortunate. Can you confirm that you have the Quicklaunch extension version 15.0.4 installed?

**Comment by lazworks** [comment]
Yes, I do have version 15.0.4 installed.

**Comment by justcla** [comment]
We identified a bug with "CollapseRegions" and "ExpandRegions". We changed the code to detect the regions properly (by checking each line to see if the content starts with "#region". Previously, regions were being detected in a different way, which stopped working a little while ago when the underlying data structure changed. Now we detect it properly, and we've tested that it works in the latest versions.
So, maybe you're talking about a different bug. (?)

Just in case we're talking about different things, can you please describe what you are doing, what you are expecting and what is happening? Thanks

**Comment by lazworks** [comment]
Create one or more regions in code.
Activate Quick Launch and type Col and then click on Collapse Regions.
Collapse state of Regions remain unchanged.

**Comment by ygoe** [comment]
Thank you for the update, it works for me now.

**Comment by jncatanz** [comment]
Collapse and Expand regions works for C# but not for Visual Basic projects. I am using Visual Studio Professional 2017 version 15.5.2 and Quick Launch Tasks 1.5.0.4.

**Comment by TechnoFrolics** [comment]
Hello,

I am failing to get Collapse and Expand regions to work in VS 2017 Community Edition 15.6.5 and 15.6.6, Quick Launch Tasks 15.0.4, native C++.

Win 7 Pro 64 bit.

**Comment by prnadago** [comment]
I just updated the extension for C++ code as well. It should work with #pragma region now. :)
